### PR TITLE
implement balance-sizes command

### PR DESCRIPTION
**Pull request by karadnik-flutterint**
command sets the same size for all windows that share common parent

### CODE REVIEW

File: Sources/Common/cmdArgs/other/CmdKind.swift [hunk 1]
```diff
@@ -31,4 +31,5 @@ public enum CmdKind: String, CaseIterable, Equatable {
     case workspaceBackAndForth = "workspace-back-and-forth"
 
     case serverVersionInternalCommand = "server-version-internal-command"
+    case balanceSizes = "balance-sizes"
```
Comment (nikitabobko) [inline Sources/Common/cmdArgs/other/CmdKind.swift:34]: Entries in this enum are sorted alphabetically

File: Sources/AppBundle/command/BalanceSizesCommand.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import AppKit
+import Common
+import Foundation
+
+struct BalanceSizesCommand: Command {
```
Comment (nikitabobko) [inline Sources/AppBundle/command/BalanceSizesCommand.swift:5]: Please create the documentation `./docs/aerospace-balance-sizes.adoc`, `./docs/commands.adoc` and update `args-grammar.conf` (it's used for shell completion)

File: Sources/AppBundle/command/BalanceSizesCommand.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        balance(state.subject.workspace.rootTilingContainer)
+        return true
+    }
+
+    func balance(_ parent: TilingContainer) {
+        if parent.layout != Layout.tiles { return }
+        parent.children.forEach { child in
+            child.setWeight(parent.orientation, 1)
+
+            if let child = child as? TilingContainer, !child.isEffectivelyEmpty {
```
Comment (nikitabobko) [inline Sources/AppBundle/command/BalanceSizesCommand.swift:19]: `child.isEffectivelyEmpty` is redundant. Effectively empty containers are automatically removed before commands get called

File: Sources/AppBundleTests/command/BalanceSizesCommandTest.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+                    TestWindow(id: 1, parent: $0).setWeight(wsp.rootTilingContainer.orientation, 1)
+                    TestWindow(id: 2, parent: $0).setWeight(wsp.rootTilingContainer.orientation, 2)
+                    TestWindow(id: 3, parent: $0).setWeight(wsp.rootTilingContainer.orientation, 3)
+                }
+            }
+
+        BalanceSizesCommand().run(CommandMutableState.init(.emptyWorkspace(name)))
+
+        workspace.rootTilingContainer.children.forEach { window in
+                XCTAssertEqual(window.getWeight(workspace.rootTilingContainer.orientation), 1)
```
Comment (nikitabobko) [inline Sources/AppBundleTests/command/BalanceSizesCommandTest.swift:20]: Indentation is incorrect

File: Sources/AppBundleTests/command/BalanceSizesCommandTest.swift [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    override func setUpWithError() throws { setUpWorkspacesForTests() }
+
+    func testBalanceSizesCommand() {
+        let workspace = Workspace.get(byName: name).apply { wsp in
+                wsp.rootTilingContainer.apply {
+                    TestWindow(id: 1, parent: $0).setWeight(wsp.rootTilingContainer.orientation, 1)
+                    TestWindow(id: 2, parent: $0).setWeight(wsp.rootTilingContainer.orientation, 2)
+                    TestWindow(id: 3, parent: $0).setWeight(wsp.rootTilingContainer.orientation, 3)
+                }
+            }
```
Comment (nikitabobko) [inline Sources/AppBundleTests/command/BalanceSizesCommandTest.swift:15]: Indentation is incorrect

File: Sources/AppBundle/command/other/parseCommand.swift [hunk 6]
```diff
@@ -81,6 +81,8 @@ extension CmdArgs {
                 command = WorkspaceCommand(args: self as! WorkspaceCmdArgs)
             case .workspaceBackAndForth:
                 command = WorkspaceBackAndForthCommand()
+            case .balanceSizes:
```
Comment (nikitabobko) [inline Sources/AppBundle/command/other/parseCommand.swift:84]: case handling should be sorted alphabetically

File: Sources/Common/cmdArgs/other/parseCmdArgs.swift [hunk 7]
```diff
@@ -82,6 +82,8 @@ private func initSubcommands() -> [String: any SubCommandParserProtocol] {
                 result[kind.rawValue] = SubCommandParser(parseWorkspaceCmdArgs)
             case .workspaceBackAndForth:
                 result[kind.rawValue] = defaultSubCommandParser(WorkspaceBackAndForthCmdArgs.init)
+            case .balanceSizes:
```
Comment (nikitabobko) [inline Sources/Common/cmdArgs/other/parseCmdArgs.swift:85]: Entries handling should be sorted alphabetically

File: Sources/AppBundle/command/BalanceSizesCommand.swift [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    let args = BalanceSizesCmdArgs(rawArgs: [])
+
+    func _run(_ state: CommandMutableState, stdin: String) -> Bool {
+        check(Thread.current.isMainThread)
+        balance(state.subject.workspace.rootTilingContainer)
+        return true
+    }
+
+    func balance(_ parent: TilingContainer) {
+        if parent.layout != Layout.tiles { return }
```
Comment (nikitabobko) [inline Sources/AppBundle/command/BalanceSizesCommand.swift:15]: You should skip `setWeight` for children of non-tiles containers (children of accordion containers), but you still should recursively check children of accordion containers

Example:
```
h_accordion
  window1
  v_tiles
    window2  <-- need to call setWeight for this leaf
    window3  <-- need to call setWeight for this leaf
```

**Review by nikitabobko — CHANGES_REQUESTED** [review]
Thanks! A few comments need to be fixed. Mainly the documentation

File: Sources/AppBundle/command/BalanceSizesCommand.swift (hunk 2 above)
Comment (nikitabobko) [reply Sources/AppBundle/command/BalanceSizesCommand.swift:5]: Please use `./build-docs.sh` to check how site and manpages look like after your changes (`./.man` and `./.site/commands.html`)

File: Sources/AppBundle/command/BalanceSizesCommand.swift (hunk 8 above)
Comment (karadnik-flutterint) [reply Sources/AppBundle/command/BalanceSizesCommand.swift:15]: Oh, i tought accordion cannot have tiles. Will filx

**Comment by karadnik-flutterint** [comment]
@nikitabobko, all comments should be addressed now.

this is how docs look like
Screenshot 2024-06-20 at 17 11 38

**Comment by nikitabobko** [comment]
Thanks!

Merged with the following patch:
```diff
diff --git a/Sources/AppBundle/command/BalanceSizesCommand.swift b/Sources/AppBundle/command/BalanceSizesCommand.swift
index d1d7c369..9456cdc2 100644
--- a/Sources/AppBundle/command/BalanceSizesCommand.swift
+++ b/Sources/AppBundle/command/BalanceSizesCommand.swift
@@ -3,7 +3,7 @@ import Common
 import Foundation
 
 struct BalanceSizesCommand: Command {
-    let args = BalanceSizesCmdArgs(rawArgs: [])
+    let args: BalanceSizesCmdArgs
 
     func _run(_ state: CommandMutableState, stdin: String) -> Bool {
         check(Thread.current.isMainThread)
@@ -12,9 +12,10 @@ struct BalanceSizesCommand: Command {
     }
 
     func balance(_ parent: TilingContainer) {
-        parent.children.forEach { child in
-            if parent.layout == .tiles {
-                child.setWeight(parent.orientation, 1)
+        for child in parent.children {
+            switch parent.layout {
+                case .tiles: child.setWeight(parent.orientation, 1)
+                case .accordion: break // Do nothing
             }
             if let child = child as? TilingContainer {
                 balance(child)
diff --git a/Sources/AppBundle/command/other/parseCommand.swift b/Sources/AppBundle/command/other/parseCommand.swift
index d58a88f9..41bdf8eb 100644
--- a/Sources/AppBundle/command/other/parseCommand.swift
+++ b/Sources/AppBundle/command/other/parseCommand.swift
@@ -20,7 +20,7 @@ extension CmdArgs {
         let command: any Command
         switch Self.info.kind {
             case .balanceSizes:
-                command = BalanceSizesCommand()
+                command = BalanceSizesCommand(args: self as! BalanceSizesCmdArgs)
             case .close:
                 command = CloseCommand(args: self as! CloseCmdArgs)
             case .closeAllWindowsButCurrent:
diff --git a/Sources/Cli/subcommandDescriptionsGenerated.swift b/Sources/Cli/subcommandDescriptionsGenerated.swift
index 335eb414..df36c278 100644
--- a/Sources/Cli/subcommandDescriptionsGenerated.swift
+++ b/Sources/Cli/subcommandDescriptionsGenerated.swift
@@ -1,5 +1,6 @@
 // FILE IS GENERATED BY generate.sh
 let subcommandDescriptions = [
+    ["  balance-sizes", "Balance sizes of all windows in the current workspace"],
     ["  close-all-windows-but-current", "On the focused workspace, close all windows but current"],
     ["  close", "Close the focused window"],
     ["  config", "Query AeroSpace config options"],
diff --git a/docs/aerospace-balance-sizes.adoc b/docs/aerospace-balance-sizes.adoc
index 3f49ee3b..41dbd4f8 100644
--- a/docs/aerospace-balance-sizes.adoc
+++ b/docs/aerospace-balance-sizes.adoc
@@ -2,24 +2,23 @@
 include::util/man-attributes.adoc[]
 :manname: aerospace-balance-sizes
 // tag::purpose[]
-:manpurpose: Balance the sizes of all opened windows in current workspace
+:manpurpose: Balance sizes of all windows in the current workspace
 // end::purpose[]
 
 == Synopsis
 // tag::synopsis[]
-aerospace balance-sizes [-h|--help] 
+aerospace balance-sizes [-h|--help]
 // end::synopsis[]
 
 == Description
-Balance the sizes of all opened windows in current workspace
 
 // tag::body[]
 {manpurpose}
 
+// end::body[]
+
 include::util/conditional-options-header.adoc[]
 
 -h, --help:: Print help
 
-// end::body[]
-
 include::util/man-footer.adoc[]
```

**Pull request closed without merging**